### ISSUE TITLE
BERT Server Prediction Doesn't Seem To Work

### ISSUE THREAD

**Issue by bonedaddy** [nlpodyssey/spago#14]
I'm attempting to use the BERT server, and have successfully gotten the `/answer` API call to work. I can't seem to find much information on `BERT prediction` in general, but I'm guessing its used to predict what the next sentence will be?

Based on this I tried sending a JSON request to the `/predict` route like so, but it gives an empty response:

```shell
$> curl -d '{"text": "BERT is a technique for NLP developed by Google. BERT was created and published in 2018 by Jacob Devlin and his colleagues from Google."}' -H Content-Type=application/json http://localhost:1987/predict
{"tokens":[]}
```

Contrast this with `discriminate` which works:

```shell
$> curl -d '{"text": "BERT is a technique for NLP developed by Google. BERT was created and published in 2018 by Jacob Devlin and his colleagues from Google."}' -H Content-Type=application/json http://localhost:1987/discriminate
{"tokens":[{"text":"BERT","start":0,"end":4,"label":"FAKE"},{"text":"is","start":5,"end":7,"label":"FAKE"},{"text":"a","start":8,"end":9,"label":"FAKE"},{"text":"technique","start":10,"end":19,"label":"FAKE"},{"text":"for","start":20,"end":23,"label":"FAKE"},{"text":"NLP","start":24,"end":27,"label":"FAKE"},{"text":"developed","start":28,"end":37,"label":"FAKE"},{"text":"by","start":38,"end":40,"label":"FAKE"},{"text":"Google","start":41,"end":47,"label":"FAKE"},{"text":".","start":47,"end":48,"label":"FAKE"},{"text":"BERT","start":49,"end":53,"label":"FAKE"},{"text":"was","start":54,"end":57,"label":"FAKE"},{"text":"created","start":58,"end":65,"label":"FAKE"},{"text":"and","start":66,"end":69,"label":"FAKE"},{"text":"published","start":70,"end":79,"label":"FAKE"},{"text":"in","start":80,"end":82,"label":"FAKE"},{"text":"2018","start":83,"end":87,"label":"FAKE"},{"text":"by","start":88,"end":90,"label":"FAKE"},{"text":"Jacob","start":91,"end":96,"label":"FAKE"},{"text":"Devlin","start":97,"end":103,"label":"FAKE"},{"text":"and","start":104,"end":107,"label":"FAKE"},{"text":"his","start":108,"end":111,"label":"FAKE"},{"text":"colleagues","start":112,"end":122,"label":"FAKE"},{"text":"from","start":123,"end":127,"label":"FAKE"},{"text":"Google","start":128,"end":134,"label":"FAKE"},{"text":".","start":134,"end":135,"label":"FAKE"}]}
```

Also, what does `discriminate` do exactly?  I found an article explaining all the details of [BERT](https://medium.com/dissecting-bert/dissecting-bert-part2-335ff2ed9c73) but cant seem to find out what `discriminate` does

**Comment by matteo-grella** [comment]
The operation of these APIs is not what you've guessed. My bad, I didn't describe APIs in the README yet.

### Predict

`predict` performs a prediction based on a trained **Masked Language Model** (MLM).
In short, MLM is a fill-in-the-blank task, where the objective is to use the context words surrounding a `[MASK]` token to try to predict what that `[MASK]` word should be.

To try the `predict` API, it is necessary that the underlying model contains all the necessary neural layers. My advice is to start with the base BERT English model trained by Hugging Face (exact name for the import: `bert-base-cased`)

Once you have imported the model and the server is listening, try this:

```
curl -k -d '{"text": "Sorry I did not get a chance to answer your [MASK] earlier."}' -H "Content-Type: application/json" "http://127.0.0.1:1987/predict?pretty"
```

It should print:

```
{
    "tokens": [
        {
            "text": "question",
            "start": 44,
            "end": 50,
            "label": "PREDICTED"
        }
    ],
    "took": 416
}
```

You can experiment with more `[MASK]` tokens and the model will generate the most likely substitution for each. Keep in mind that the more tokens are masked the less context is usable and therefore the accuracy may drop.

You can even mix several languages in the same sentence using a multi-lingual model (exact name for the import: `bert-base-multilingual-cased`).

For example:

```console
curl -k -d '{"text": "Io sono italiano quindi parlo [MASK] , but as soon as I am with my German colleagues I switch to [MASK] ."}' -H "Content-Type: application/json" "http://127.0.0.1:1987/predict?pretty"
```

It should print:

```json
{
    "tokens": [
        {
            "text": "italiano",
            "start": 30,
            "end": 36,
            "label": "PREDICTED"
        },
        {
            "text": "English",
            "start": 97,
            "end": 103,
            "label": "PREDICTED"
        }
    ],
    "took": 469
}
````

Cool! Isn't it?

### Discriminate

There's something in the tests I'm doing that doesn't feel right. I'll come back to this later.

**Comment by bonedaddy** [comment]
Woah that's really cool :O

**Comment by bonedaddy** [comment]
I'm trying out the prediction now and for some reason am getting a result like this:

```
&{Tokens:[{Text:[PAD] Start:44 End:50 Label:PREDICTED}] Took:390}
```

Text I tried predicting:
```
Sorry I did not get a chance to answer your [MASK] earlier.
```

model used: `deepset/bert-base-cased-squad2`

**Comment by matteo-grella** [comment]
This is because the model `deepset/bert-base-cased-sqad2` is **fine-tuned** for question-answering on the [SQuAD2.0](https://rajpurkar.github.io/SQuAD-explorer/) dataset.

As a consequence of fine-tuning, some layers used during the so-called **pre-training** phase in which the BERT model is trained to learn language patterns (or in other words to become a [super-parrot](https://medium.com/@ElementalCognition/can-super-parrots-ever-achieve-language-understanding-8307dfd3e87c)) are discarded.

The masked token prediction works on those layers that are no longer available in the fine-tuned SQuAD model:

```go
// BERT Model
type Model struct {
	Config          Config
	Vocabulary      *vocabulary.Vocabulary
	Embeddings      *Embeddings
	Encoder         *Encoder
	Predictor       *Predictor // <---- This is the layer only available with basic "pre-trained"
	Discriminator   *Discriminator 
	Pooler          *Pooler
	SeqRelationship *linear.Model
	SpanClassifier  *SpanClassifier
}
```

As I said in my first answer, I suggest to use the following models:

- `bert-base-cased`

- `bert-base-multilingual-cased`

These are the exact names of the models to import, without subdirectories as for [deepset](https://deepset.ai/) models.

Please note that the server CLI has changed here: #28

**Comment by matteo-grella** [comment]
It's working like it's supposed to.
